File: Examples/android/CMeSdkJava/app/src/main/java/me/connect/sdk/java/sample/SingleLiveData.java
package me.connect.sdk.java.sample;

import androidx.annotation.MainThread;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * {@link MutableLiveData} variant that sends data only once.
 * @param <T>
 */
public class SingleLiveData<T> extends MutableLiveData<T> {

    private final AtomicBoolean pending = new AtomicBoolean(false);

    @MainThread
    public void observeOnce(LifecycleOwner lifecycleOwner, final Observer<T> observer) {
        super.observe(lifecycleOwner, t -> {
            if (pending.compareAndSet(true, false)) {
                observer.onChanged(t);
            }
        });

    }

    @MainThread
    public void setValue(T t) {
        pending.set(true);
        super.setValue(t);
    }
}

File: Examples/android/CMeSdkJava/app_kt/src/main/kotlin/me/connect/sdk/java/samplekt/SingleLiveData.kt
package me.connect.sdk.java.samplekt

import androidx.annotation.MainThread
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import java.util.concurrent.atomic.AtomicBoolean


/**
 * [MutableLiveData] variant that sends data only once.
 * @param <T>
</T> */
class SingleLiveData<T> : MutableLiveData<T>() {
    private val pending = AtomicBoolean(false)

    @MainThread
    fun observeOnce(lifecycleOwner: LifecycleOwner, observer: Observer<T>) {
        super.observe(lifecycleOwner, Observer { t: T ->
            if (pending.compareAndSet(true, false)) {
                observer.onChanged(t)
            }
        })
    }

    @MainThread
    override fun setValue(t: T) {
        pending.set(true)
        super.setValue(t)
    }
}
